pjsip/src/pjsip-simple/dlg_event.c

```diff
@@ -0,0 +1,604 @@
+ *        is not PJ_SUCCESS, res_hdr may be added with Warning header.
+ */
+static pj_status_t
+dlg_event_process_rx_notify(pjsip_dlg_event *dlgev,
+                            pjsip_rx_data *rdata,
+                            int *p_st_code,
+                            pj_str_t **p_st_text,
+                            pjsip_hdr *res_hdr)
+{
+    pjsip_ctype_hdr *ctype_hdr;
+    pj_status_t status = PJ_SUCCESS;
+
+    *p_st_text = NULL;
+
+    /* Check Content-Type and msg body are dlgevent. */
+    ctype_hdr = rdata->msg_info.ctype;
+
+    if (ctype_hdr==NULL || rdata->msg_info.msg->body==NULL)
+    {
+        pjsip_warning_hdr *warn_hdr;
+        pj_str_t warn_text;
+
+        *p_st_code = PJSIP_SC_BAD_REQUEST;
+
+        warn_text = pj_str("Message body is not dlgevent");
+        warn_hdr = pjsip_warning_hdr_create(rdata->tp_info.pool, 399,
+                            pjsip_endpt_name(dlgev->dlg->endpt),
+                            &warn_text);
+        pj_list_push_back(res_hdr, warn_hdr);
+
+        return PJSIP_ERRNO_FROM_SIP_STATUS(PJSIP_SC_BAD_REQUEST);
+    }
+
+    /* Parse content. */
+    if (pj_stricmp(&ctype_hdr->media.type, &STR_DIALOG_APPLICATION)==0 &&
+        pj_stricmp(&ctype_hdr->media.subtype, &STR_DIALOG_INFO_XML)==0)
+    {
+        status = pjsip_dlg_event_parse_dialog_info(rdata, dlgev->tmp_pool,
+                                                   &dlgev->tmp_status);
+    }
+    else {
+        status = PJSIP_SIMPLE_EBADCONTENT;
+    }
+
+    if (status != PJ_SUCCESS) {
+    /* Unsupported or bad Content-Type */
+        if (PJSIP_DLG_EVENT_BAD_CONTENT_RESPONSE >= 300) {
+            pjsip_accept_hdr *accept_hdr;
+            pjsip_warning_hdr *warn_hdr;
+
+            *p_st_code = PJSIP_DLG_EVENT_BAD_CONTENT_RESPONSE;
+
+            /* Add Accept header */
+            accept_hdr = pjsip_accept_hdr_create(rdata->tp_info.pool);
+            accept_hdr->values[accept_hdr->count++] = STR_APP_DIALOG_INFO_XML;
+            pj_list_push_back(res_hdr, accept_hdr);
+
+            /* Add Warning header */
+            warn_hdr = pjsip_warning_hdr_create_from_status(
+                        rdata->tp_info.pool,
+                        pjsip_endpt_name(dlgev->dlg->endpt),
+                        status);
+            pj_list_push_back(res_hdr, warn_hdr);
+
+            return status;
+        } else {
+            pj_assert(PJSIP_DLG_EVENT_BAD_CONTENT_RESPONSE/100 == 2);
+            PJ_PERROR(4,(THIS_FILE, status,
+                      "Ignoring dlgev error due to "
+                      "PJSIP_DLG_EVENT_BAD_CONTENT_RESPONSE setting [%d]",
+                      PJSIP_DLG_EVENT_BAD_CONTENT_RESPONSE));
+            *p_st_code = PJSIP_DLG_EVENT_BAD_CONTENT_RESPONSE;
+            status = PJ_SUCCESS;
+        }
+    }
+
+    /* If application calls dlgev_get_status(), redirect the call to
+     * retrieve the temporary status.
+     */
+    dlgev->tmp_status._is_valid = PJ_TRUE;
```

Comment:
This `_is_valid` should not set to TRUE when `pjsip_dlg_event_parse_dialog_info()` return non-success? As otherwise `pjsip_dlg_event_get_status()` will copy status from `tmp_status` which may contain invalid data (as parsing failed)?